File: packages/core_crypto/ios/core_crypto/Sources/core_crypto/EvmNetworkPolicy.swift
import Foundation

enum EvmNetworkPolicy {
  private static let domains: [UInt64: String] = [
    1: "eth", 11155111: "eth", 137: "polygon", 80002: "polygon",
    8453: "base", 84532: "base", 42161: "arbitrum", 421614: "arbitrum",
    43114: "avalanche", 43113: "avalanche", 56: "bnb", 97: "bnb",
  ]
  static func allows(coin: String, raw: Data, allowUnknown: Bool) -> Bool {
    let significant = raw.drop(while: { $0 == 0 })
    guard !significant.isEmpty else { return false }
    guard significant.count <= 8 else { return allowUnknown }
    let id = significant.reduce(UInt64(0)) { ($0 << 8) | UInt64($1) }
    guard let expected = domains[id] else { return allowUnknown }
    return expected == coin
  }
}

File: packages/core_crypto/android/src/main/kotlin/com/ktwallet/core_crypto/EvmNetworkPolicy.kt
package com.ktwallet.core_crypto

import java.math.BigInteger

internal object EvmNetworkPolicy {
    private val domains = mapOf(
        1L to "eth", 11155111L to "eth", 137L to "polygon", 80002L to "polygon",
        8453L to "base", 84532L to "base", 42161L to "arbitrum", 421614L to "arbitrum",
        43114L to "avalanche", 43113L to "avalanche", 56L to "bnb", 97L to "bnb",
    )
    fun allows(coin: String, raw: ByteArray, allowUnknown: Boolean): Boolean {
        if (raw.isEmpty()) return false
        val id = BigInteger(1, raw)
        if (id.signum() <= 0) return false
        val expected = if (id.bitLength() <= 63) domains[id.toLong()] else null
        return if (expected == null) allowUnknown else expected == coin
    }
}
